In [1]:
import numpy as np 
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt

# Q1

Problem 1 description: In the first question, for 6000 sets of labeled data, we need to build a binary classification model, and then use this training model to classify 100 sets of unlabeled data. In this question, we need to divide the training set and test set by ourselves, and strive to find the optimal classification model.

Firstly, the training data given in the first question was opened and there were 6000 entries, each with 129 columns. After observing that the first 128 columns were features and the last one marked whether it was under attack or not, I counted the summation of the last column and it was exactly 3000, which matched the description of the question (half were under attack and the other half were not).

In [2]:
data=pd.read_csv("TrainingDataBinary.csv",header=None)
data

0            1          2            3           4    \
0      70.399324  127673.0908 -49.572308  127648.0176 -169.578319   
1      73.688102  130280.7109 -46.300719  130255.6377 -166.278082   
2      73.733939  130305.7842 -46.254883  130280.7109 -166.232245   
3      74.083443  130581.5902 -45.899649  130556.5169 -165.882741   
4      74.553268  131083.0556 -45.424094  131057.9823 -165.424375   
...          ...          ...        ...          ...         ...   
5995  116.889120  131860.3269  -3.076783  131810.1804 -123.094253   
5996  116.849013  131810.1804  -3.116890  131760.0339 -123.128630   
5997  116.384917  131734.9606  -3.586716  131684.8140 -123.586996   
5998  111.125164  130506.3704  -8.846468  130456.2238 -128.858208   
5999  110.878793  130481.2971  -9.092840  130456.2238 -129.104580   

              5           6          7          8          9    ...  119  120  \
0     127723.2374   65.689611  605.91099 -57.003571  626.78553  ...    0    0   
1     130355.9307   71.831719  483.59351 -50.947407  500.98896  ...    0    0   
2     130381.0040   71.808800  483.59351 -50.913030  500.98896  ...    0    0   
3     130656.8100   72.152575  482.86107 -50.437475  499.15786  ...    0    0   
4     131158.2754   72.118198  484.50906 -50.013486  497.69298  ...    0    0   
...           ...         ...        ...        ...        ...  ...  ...  ...   
5995  131910.4735  114.780635  376.10794  -5.254023  374.82617  ...    0    0   
5996  131885.4002  114.769176  376.29105  -5.322778  374.82617  ...    0    0   
5997  131785.1071  114.299351  376.47416  -5.849899  374.82617  ...    0    0   
5998  130556.5169  106.667553  478.83265 -13.464508  477.73399  ...    0    0   
5999  130556.5169  106.392533  478.83265 -13.750987  477.91710  ...    0    0   

      121  122  123  124  125  126  127  128  
0       0    0    0    0    0    0    0    0  
1       0    0    0    0    0    0    0    0  
2       0    0    0    0    0    0    0    0  
3       0    0    0    0    0    0    0    0  
4       0    0    0    0    0    0    0    0  
...   ...  ...  ...  ...  ...  ...  ...  ...  
5995    0    0    0    0    0    0    0    0  
5996    0    0    0    0    0    0    0    0  
5997    0    0    0    0    0    0    0    0  
5998    0    0    0    0    0    0    0    0  
5999    0    0    0    0    0    0    0    0  

[6000 rows x 129 columns]

In [3]:
from sklearn import preprocessing
X = data.iloc[:,0:128].to_numpy()
y = data.iloc[:,-1].to_numpy()
X.shape

(6000, 128)

In the next step we open the test set, which is not labelled as being attacked, so we need to manually split the training and validation sets in the training data and use the validation set to evaluate the accuracy of the chosen model.

In [4]:
test=pd.read_csv("TestingDataBinary.csv",header=None)
X_test = test.iloc[:,:].to_numpy()
test

0            1           2            3           4            5    \
0    12.118057  131810.1804 -107.847846  131183.3486  132.146986  131860.3269   
1    12.049302  131810.1804 -107.922330  131183.3486  132.066772  131860.3269   
2    12.043573  131810.1804 -107.928060  131183.3486  132.061042  131860.3269   
3    12.037843  131810.1804 -107.939519  131183.3486  132.061042  131860.3269   
4    11.946170  131810.1804 -108.031192  131183.3486  131.963639  131860.3269   
..         ...          ...         ...          ...         ...          ...   
95 -173.606212  132161.2062   66.428727  132136.1329  -53.577283  132236.4260   
96 -172.070685  132612.5250   67.969983  132587.4518  -52.047486  132687.7448   
97 -172.013389  132537.3052   68.027279  132512.2320  -51.984461  132612.5250   
98 -172.076415  132411.9389   67.969983  132386.8656  -52.041757  132462.0854   
99 -172.168088  132361.7923   67.872580  132336.7191  -52.139159  132411.9389   

           6          7           8          9    ...  118  119  120  121  \
0    10.089787  370.79775 -109.664122  369.88220  ...    0    0    0    0   
1     9.986654  370.79775 -109.899035  370.61464  ...    0    0    0    0   
2     9.958006  370.79775 -109.887576  370.61464  ...    0    0    0    0   
3     9.980925  370.79775 -109.870387  370.61464  ...    0    0    0    0   
4     9.860604  370.79775 -110.059463  370.79775  ...    0    0    0    0   
..         ...        ...         ...        ...  ...  ...  ...  ...  ...   
95 -172.826989  259.28376   67.282434  261.66419  ...    0    0    0    0   
96 -171.199789  258.18510   69.047144  260.56553  ...    0    0    0    0   
97 -171.159682  258.73443   69.092981  260.56553  ...    0    0    0    0   
98 -171.257085  258.91754   69.098710  260.56553  ...    0    0    0    0   
99 -171.337299  259.28376   69.007037  260.56553  ...    0    0    0    0   

    122  123  124  125  126  127  
0     0    0    0    0    0    0  
1     0    0    0    0    0    0  
2     0    0    0    0    0    0  
3     0    0    0    0    0    0  
4     0    0    0    0    0    0  
..  ...  ...  ...  ...  ...  ...  
95    0    0    0    0    0    0  
96    0    0    0    0    0    0  
97    0    0    0    0    0    0  
98    0    0    0    0    0    0  
99    0    0    0    0    0    0  

[100 rows x 128 columns]

Three methods were chosen to classify the training set: logistic regression, random forest and SVR, each with StratifiedShuffle() for stratified sampling, which resulted in a more uniform distribution of training and validation data and a more generalised model. We split the data into five groups and for each group an accuracy-based score is given as follows.

Method 1 Logistic Regression

In [17]:
# from sklearn.model_selection import StratifiedShuffleSplit
# from sklearn.linear_model import LogisticRegression
# from sklearn.metrics import accuracy_score

# split = StratifiedShuffleSplit(n_splits = 5,test_size = 0.2,random_state = 42)
# LR = LogisticRegression()
# scores = []

# for train_index,valid_index in split.split(X,y):
#     X_train, X_valid = X[train_index], X[valid_index]
#     y_train, y_valid = y[train_index], y[valid_index]
#     LR.fit(X_train, y_train)
#     pred = LR.predict(X_valid)
#     scores.append(accuracy_score(y_valid, pred))
 
# # get accuracy of each prediction
# print(scores)

Chosen Method: Random Forest

In [6]:
from sklearn.ensemble import RandomForestClassifier

split = StratifiedShuffleSplit(n_splits = 5,test_size = 0.2,random_state = 42)
model2=RandomForestClassifier(n_estimators=150,oob_score=False,bootstrap=False,max_features=53)
scores = []

for train_index,valid_index in split.split(X,y):
    X_train, X_valid = X[train_index], X[valid_index]
    y_train, y_valid = y[train_index], y[valid_index]
    model2.fit(X_train, y_train)
    pred = model2.predict(X_valid)
    scores.append(accuracy_score(y_valid, pred))
 
# get accuracy of each prediction
print(scores)

[0.985, 0.9833333333333333, 0.985, 0.9858333333333333, 0.985]


At last, Random Forest Classifier is chosen as the final model. The random forest algorithm uses the CART decision tree as a weak learner, and has made improvements on this basis. The algorithm repeatedly randomly draws k samples from the original training sample set N with replacement to generate a new training sample set, and then generates k classification trees to form a random forest based on the self-service sample set. The classification results of new data are voted according to the classification tree Depends on the score formed. Its essence is an improvement to the decision tree algorithm. Multiple decision trees are merged together. The establishment of each tree depends on an independently drawn sample. Each tree in the forest has the same distribution, and the classification error depends on each decision tree. The classification power of a tree and the correlations between them. Feature selection uses a random method to split each node, and then compares the errors generated in different situations. The intrinsic estimation error, classification power, and relevance that can be detected determine the number of features to choose. The classification ability of a single tree may be very small, but after a large number of decision trees are randomly generated, a test sample can select the most probable classification through statistics of the classification results of each tree.


Training accuracies for each time are shown above, which are 98.5833%, 98%, 98.3333%, 98.5833% and 98.25%. 

In [7]:
y_test=model2.predict(X_test)
print(y_test)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1
 1 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]


In [8]:
y1=pd.DataFrame(y_test)
df = pd.concat([test,y1],axis=1)
df

0            1           2            3           4            5    \
0    12.118057  131810.1804 -107.847846  131183.3486  132.146986  131860.3269   
1    12.049302  131810.1804 -107.922330  131183.3486  132.066772  131860.3269   
2    12.043573  131810.1804 -107.928060  131183.3486  132.061042  131860.3269   
3    12.037843  131810.1804 -107.939519  131183.3486  132.061042  131860.3269   
4    11.946170  131810.1804 -108.031192  131183.3486  131.963639  131860.3269   
..         ...          ...         ...          ...         ...          ...   
95 -173.606212  132161.2062   66.428727  132136.1329  -53.577283  132236.4260   
96 -172.070685  132612.5250   67.969983  132587.4518  -52.047486  132687.7448   
97 -172.013389  132537.3052   68.027279  132512.2320  -51.984461  132612.5250   
98 -172.076415  132411.9389   67.969983  132386.8656  -52.041757  132462.0854   
99 -172.168088  132361.7923   67.872580  132336.7191  -52.139159  132411.9389   

           6          7           8          9    ...  119  120  121  122  \
0    10.089787  370.79775 -109.664122  369.88220  ...    0    0    0    0   
1     9.986654  370.79775 -109.899035  370.61464  ...    0    0    0    0   
2     9.958006  370.79775 -109.887576  370.61464  ...    0    0    0    0   
3     9.980925  370.79775 -109.870387  370.61464  ...    0    0    0    0   
4     9.860604  370.79775 -110.059463  370.79775  ...    0    0    0    0   
..         ...        ...         ...        ...  ...  ...  ...  ...  ...   
95 -172.826989  259.28376   67.282434  261.66419  ...    0    0    0    0   
96 -171.199789  258.18510   69.047144  260.56553  ...    0    0    0    0   
97 -171.159682  258.73443   69.092981  260.56553  ...    0    0    0    0   
98 -171.257085  258.91754   69.098710  260.56553  ...    0    0    0    0   
99 -171.337299  259.28376   69.007037  260.56553  ...    0    0    0    0   

    123  124  125  126  127  0    
0     0    0    0    0    0    1  
1     0    0    0    0    0    1  
2     0    0    0    0    0    1  
3     0    0    0    0    0    1  
4     0    0    0    0    0    1  
..  ...  ...  ...  ...  ...  ...  
95    0    0    0    0    0    0  
96    0    0    0    0    0    0  
97    0    0    0    0    0    0  
98    0    0    0    0    0    0  
99    0    0    0    0    0    0  

[100 rows x 129 columns]

In [9]:
df.to_csv('TestingResultsBinary.csv')

# Q2

Description of the problem 2: In the second question, for 6000 sets of labeled data, we need to build a multi-classification model, and then use this training model to classify 100 sets of unlabeled data. In this question, we need to divide the training set and test set by ourselves, and strive to find the optimal classification model.

Firstly, the training data given in the first question was opened and there were 6000 entries, each with 129 columns. After observing that the first 128 columns were features and the last one is the label, I counted the summation of the last column and it was exactly 3000, which matched the description of the question (half were under attack and the other half were not). The subsequent process is similar to Question 1.

In [10]:
data2=pd.read_csv("TrainingDataMulti.csv",header=None)
data2

0            1          2            3           4    \
0      70.399324  127673.0908 -49.572308  127648.0176 -169.578319   
1      73.688102  130280.7109 -46.300719  130255.6377 -166.278082   
2      73.733939  130305.7842 -46.254883  130280.7109 -166.232245   
3      74.083443  130581.5902 -45.899649  130556.5169 -165.882741   
4      74.553268  131083.0556 -45.424094  131057.9823 -165.424375   
...          ...          ...        ...          ...         ...   
5995  116.889120  131860.3269  -3.076783  131810.1804 -123.094253   
5996  116.849013  131810.1804  -3.116890  131760.0339 -123.128630   
5997  116.384917  131734.9606  -3.586716  131684.8140 -123.586996   
5998  111.125164  130506.3704  -8.846468  130456.2238 -128.858208   
5999  110.878793  130481.2971  -9.092840  130456.2238 -129.104580   

              5           6          7          8          9    ...  119  120  \
0     127723.2374   65.689611  605.91099 -57.003571  626.78553  ...    0    0   
1     130355.9307   71.831719  483.59351 -50.947407  500.98896  ...    0    0   
2     130381.0040   71.808800  483.59351 -50.913030  500.98896  ...    0    0   
3     130656.8100   72.152575  482.86107 -50.437475  499.15786  ...    0    0   
4     131158.2754   72.118198  484.50906 -50.013486  497.69298  ...    0    0   
...           ...         ...        ...        ...        ...  ...  ...  ...   
5995  131910.4735  114.780635  376.10794  -5.254023  374.82617  ...    0    0   
5996  131885.4002  114.769176  376.29105  -5.322778  374.82617  ...    0    0   
5997  131785.1071  114.299351  376.47416  -5.849899  374.82617  ...    0    0   
5998  130556.5169  106.667553  478.83265 -13.464508  477.73399  ...    0    0   
5999  130556.5169  106.392533  478.83265 -13.750987  477.91710  ...    0    0   

      121  122  123  124  125  126  127  128  
0       0    0    0    0    0    0    0    0  
1       0    0    0    0    0    0    0    0  
2       0    0    0    0    0    0    0    0  
3       0    0    0    0    0    0    0    0  
4       0    0    0    0    0    0    0    0  
...   ...  ...  ...  ...  ...  ...  ...  ...  
5995    0    0    0    0    0    0    0    0  
5996    0    0    0    0    0    0    0    0  
5997    0    0    0    0    0    0    0    0  
5998    0    0    0    0    0    0    0    0  
5999    0    0    0    0    0    0    0    0  

[6000 rows x 129 columns]

In [11]:
X2 = data2.iloc[:,0:128].to_numpy()
y2 = data2.iloc[:,-1].to_numpy()
X2.shape

(6000, 128)

In [12]:
test2=pd.read_csv("TestingDataMulti.csv",header=None)
X2_test = test2.iloc[:,:].to_numpy()
test2

0            1           2            3          4            5    \
0  -100.141563  132336.7191  139.881916  132311.6458  19.875906  132411.9389   
1  -100.267614  132311.6458  139.732947  132311.6458  19.738396  132411.9389   
2  -100.301992  132336.7191  139.710029  132286.5725  19.726937  132411.9389   
3  -100.468149  132336.7191  139.555330  132286.5725  19.560779  132411.9389   
4  -100.479609  132336.7191  139.538141  132311.6458  19.537861  132411.9389   
..         ...          ...         ...          ...        ...          ...   
95 -179.118066  131584.5210   60.905414  131559.4477 -59.106326  131659.7408   
96  179.060770  131258.5685   59.083408  131208.4219 -60.922602  131333.7883   
97  178.642511  131208.4219   58.670878  131183.3486 -61.340862  131283.6417   
98  178.625322  131208.4219   58.647960  131208.4219 -61.363780  131308.7150   
99  178.619593  131233.4952   58.642230  131208.4219 -61.363780  131308.7150   

           6          7           8          9    ...  118  119  120  121  \
0   -90.590357  231.45104  142.431578  254.33979  ...    0    0    0    0   
1   -92.280582  236.94434  141.944564  255.43845  ...    0    0    0    0   
2   -92.589980  237.67678  141.881539  255.43845  ...    0    0    0    0   
3   -94.331771  242.98697  141.222637  256.53711  ...    0    0    0    0   
4   -94.417715  243.35319  141.171071  256.35400  ...    0    0    0    0   
..         ...        ...         ...        ...  ...  ...  ...  ...  ...   
95  178.430517  394.41894   58.046354  394.78516  ...    0    0    0    0   
96  175.714697  430.30850   55.445126  430.49161  ...    0    0    0    0   
97  175.290708  429.75917   54.946653  430.85783  ...    0    0    0    0   
98  175.256330  429.94228   54.923734  430.49161  ...    0    0    0    0   
99  175.262060  429.75917   54.935193  430.85783  ...    0    0    0    0   

    122  123  124  125  126  127  
0     0    0    0    0    0    0  
1     0    0    0    0    0    0  
2     0    0    0    0    0    0  
3     0    0    0    0    0    0  
4     0    0    0    0    0    0  
..  ...  ...  ...  ...  ...  ...  
95    0    0    0    0    0    0  
96    0    0    0    0    0    0  
97    0    0    0    0    0    0  
98    0    0    0    0    0    0  
99    0    0    0    0    0    0  

[100 rows x 128 columns]

At last, Random Forest Classifier is chosen as the final model. The random forest algorithm uses the CART decision tree as a weak learner, and has made improvements on this basis. The algorithm repeatedly randomly draws k samples from the original training sample set N with replacement to generate a new training sample set, and then generates k classification trees to form a random forest based on the self-service sample set. The classification results of new data are voted according to the classification tree Depends on the score formed. Its essence is an improvement to the decision tree algorithm. Multiple decision trees are merged together. The establishment of each tree depends on an independently drawn sample. Each tree in the forest has the same distribution, and the classification error depends on each decision tree. The classification power of a tree and the correlations between them. Feature selection uses a random method to split each node, and then compares the errors generated in different situations. The intrinsic estimation error, classification power, and relevance that can be detected determine the number of features to choose. The classification ability of a single tree may be very small, but after a large number of decision trees are randomly generated, a test sample can select the most probable classification through statistics of the classification results of each tree.

In [13]:
from sklearn.ensemble import RandomForestClassifier

split = StratifiedShuffleSplit(n_splits = 5,test_size = 0.2,random_state = 42)
model3=RandomForestClassifier(n_estimators=150,oob_score=False,bootstrap=False,max_features=60)
scores = []

for train_index,valid_index in split.split(X,y):
    X_train, X_valid = X2[train_index], X2[valid_index]
    y_train, y_valid = y2[train_index], y2[valid_index]
    model3.fit(X_train, y_train)
    pred = model3.predict(X_valid)
    scores.append(accuracy_score(y_valid, pred))
 
# get accuracy of each prediction
print(scores)

[0.9433333333333334, 0.9433333333333334, 0.9408333333333333, 0.9541666666666667, 0.9508333333333333]


Training accuracies for each time are shown above, which are 94.0833%, 94.08333%, 93.5833%, 95.5% and 95%. 

In [14]:
y2_test=model3.predict(X2_test)
print(y2_test)

[2 2 2 2 2 2 1 1 2 2 2 1 1 1 1 1 2 2 1 1 1 2 2 2 2 2 2 0 0 0 0 0 0 1 1 1 1
 1 1 1 1 1 2 2 2 2 2 2 2 2 1 1 2 1 2 2 2 1 1 2 1 1 1 1 1 1 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 0 0 0 0 0 0 0 0]


In [15]:
y2=pd.DataFrame(y2_test)
df = pd.concat([test2,y2],axis=1)
df

0            1           2            3          4            5    \
0  -100.141563  132336.7191  139.881916  132311.6458  19.875906  132411.9389   
1  -100.267614  132311.6458  139.732947  132311.6458  19.738396  132411.9389   
2  -100.301992  132336.7191  139.710029  132286.5725  19.726937  132411.9389   
3  -100.468149  132336.7191  139.555330  132286.5725  19.560779  132411.9389   
4  -100.479609  132336.7191  139.538141  132311.6458  19.537861  132411.9389   
..         ...          ...         ...          ...        ...          ...   
95 -179.118066  131584.5210   60.905414  131559.4477 -59.106326  131659.7408   
96  179.060770  131258.5685   59.083408  131208.4219 -60.922602  131333.7883   
97  178.642511  131208.4219   58.670878  131183.3486 -61.340862  131283.6417   
98  178.625322  131208.4219   58.647960  131208.4219 -61.363780  131308.7150   
99  178.619593  131233.4952   58.642230  131208.4219 -61.363780  131308.7150   

           6          7           8          9    ...  119  120  121  122  \
0   -90.590357  231.45104  142.431578  254.33979  ...    0    0    0    0   
1   -92.280582  236.94434  141.944564  255.43845  ...    0    0    0    0   
2   -92.589980  237.67678  141.881539  255.43845  ...    0    0    0    0   
3   -94.331771  242.98697  141.222637  256.53711  ...    0    0    0    0   
4   -94.417715  243.35319  141.171071  256.35400  ...    0    0    0    0   
..         ...        ...         ...        ...  ...  ...  ...  ...  ...   
95  178.430517  394.41894   58.046354  394.78516  ...    0    0    0    0   
96  175.714697  430.30850   55.445126  430.49161  ...    0    0    0    0   
97  175.290708  429.75917   54.946653  430.85783  ...    0    0    0    0   
98  175.256330  429.94228   54.923734  430.49161  ...    0    0    0    0   
99  175.262060  429.75917   54.935193  430.85783  ...    0    0    0    0   

    123  124  125  126  127  0    
0     0    0    0    0    0    2  
1     0    0    0    0    0    2  
2     0    0    0    0    0    2  
3     0    0    0    0    0    2  
4     0    0    0    0    0    2  
..  ...  ...  ...  ...  ...  ...  
95    0    0    0    0    0    0  
96    0    0    0    0    0    0  
97    0    0    0    0    0    0  
98    0    0    0    0    0    0  
99    0    0    0    0    0    0  

[100 rows x 129 columns]

In [16]:
df.to_csv('TestingResultsMulti.csv')